backend/server/routers/ctf.py

```diff
@@ -0,0 +1,283 @@
+    extended_courses = {
+        "COMP3821",
+        "COMP3891",
+        "COMP6841",
+        "COMP6843"
+        "COMP6845"
+    }
+    return len(extended_courses & all_courses(data)) >= 3
+
+
+def summer_course(data: PlannerData) -> bool:
+    """
+    Must take atleast one summer course
+    """
+    return any(
+        course.startswith("COMP")
+        for year in data.plan
+        for course in year[0]
+    )
+
+
+
+def term_sums_even(data: PlannerData) -> bool:
+    """
+    Check that the sum of the course codes in even terms is even
+    """
+    is_even: Callable[[int], bool] = lambda x: x % 2 == 0
+    print("Checking even")
+    for y, year in enumerate(data.plan):
+        # Exclude summer term + odd terms
+        for i, term in enumerate(year[2::2], 2):
+            term_sum = sum(map(get_code, term.keys()))
+            print(f"{y}T{i} sum: {term_sum}")
+            if not is_even(term_sum):
+                print("failed: ", term)
+                return False
+
+    return True
+
+# TODO
+def term_sums_odd(data: PlannerData) -> bool:
+    """
+    Check that the sum of the course codes in odd terms is odd
+    """
+    is_odd: Callable[[int], bool] = lambda x: x % 2 == 1
+    print("Checking odd")
+    for year in data.plan[::2]:
+        # Exclude summer term + even terms
+        for term in year[1::2]:
+            term_sum = sum(map(get_code, term.keys()))
+            if not is_odd(term_sum):
+                print("failed: ", term)
+                return False
+    return True
+ 
+def comp1511_marks(data: PlannerData) -> bool:
+    """
+    Ollie must achieve a mark of 100 in COMP1511 to keep his scholarship
+    """
+    for year in data.plan:
+        for term in year:
+            for course in term:
+                _, marks = term[course]  # type: ignore
+                if course == "COMP1511":
+                    return marks == 100
+
+    return False
+
+
+def gen_ed_sum(data: PlannerData) -> bool:
+    """
+    The sum of GENED course codes must not exceed 2200
+    """
+    return sum(map(get_code, gen_eds(all_courses(data)))) <= 2200
+
+
+def gen_ed_faculty(data: PlannerData) -> bool:
+    """
+    Gen-Eds must all be from different faculties
+    """
+    gen_eds_facs = list(map(get_faculty, gen_eds(all_courses(data))))
+    return len(gen_eds_facs) == len(set(gen_eds_facs))
+
+
+def same_code_diff_faculty(data: PlannerData) -> bool:
+    """
+    Must take two courses with the same code but, from different faculties
+    """
+    codes = list(map(get_code, all_courses(data)))
+    # Can't have duplicate of a course since it's a set
+    return len(codes) != len(set(codes))
+
+
+def math_limit(data: PlannerData) -> bool:
+    """
+    In your N-th year, you can only take N + 1 math courses
+    """
+    for i, year in enumerate(data.plan, 1):
+        num_math = len([
+            course
+            for term in year
+            for course in term
+            if course.startswith("MATH")
+        ])
```

Comment:
thoughts on using `sum` with a generator comprehension where every item is `1` instead of `course` so we don't need to build an entire list?

Reply:
actly not bad idea